feat(jwt) add support to a custom header name

raelmax opened the pull request:
### Summary

It allow the jwt plugin to work together with another authentication
method that uses the 'Authorization' header

Reference: https://discuss.konghq.com/t/custom-header-on-jwt-plugin/1524

### Full changelog

* Add a `header_names` field to jwt plugin schema;
* Change the `retrieve_token` method to  use the `header_names` field;
* Add a test to check if the custom `header_names` is used;
* Bump jwt plugin version from `0.2.0` to `0.2.1`.

Commit:
feat(jwt) add support to custom header names

It allow the jwt plugin to work together with another authentication
method that uses the 'Authorization' header

Reference: https://discuss.konghq.com/t/custom-header-on-jwt-plugin/1524

Tieske commented on kong/plugins/jwt/schema.lua:
```diff
@@ -28,6 +28,7 @@ return {
     anonymous = {type = "string", default = "", func = check_user},
     run_on_preflight = {type = "boolean", default = true},
     maximum_expiration = {type = "number", default = 0, func = check_positive},
+    header_name = {type = "string", default = "authorization"},
```
both the `uri_param_names` and the `cookie_names` accept an array. So it makes sense to also add `header_names` as an array

Tieske commented on kong/plugins/jwt/handler.lua:
```diff
@@ -13,11 +13,11 @@ local get_method     = ngx.req.get_method
 --- Retrieve a JWT in a request.
 -- Checks for the JWT in URI parameters, then in cookies, and finally
--- in the `Authorization` header.
+-- in the configured header_name(defaults to `Authorization`).
```
small typo, space before `(`

Tieske commented on kong/plugins/jwt/handler.lua:
```diff
@@ -39,7 +39,7 @@ local function retrieve_token(request, conf)
   end
 
-  local authorization_header = request.get_headers()["authorization"]
+  local authorization_header = request.get_headers()[conf.header_name or "authorization"]
```
please rebase on `next` instead of `master`, and then skip setting the default here, since those will be automatically set [since this pr](https://github.com/Kong/kong/pull/3756).

Tieske reviewed (changes requested):
Thx for your contribution @raelmax. Comments are some minor things, and a change to an array.

raelmax commented:
@Tieske I've fixed the minor changes and change the field to an array. Since I've rebased this branch on `next`, should I reopen this pull request to the `next` branch?

Tieske commented:
@raelmax at the top, right of the PR title is an `edit` button. It allows you to change the target branch. I already changed it to `next`.

Tieske commented on kong/plugins/jwt/handler.lua:
```diff
@@ -38,20 +38,25 @@ local function retrieve_token(request, conf)
+  local request_headers = request.get_headers()
+  for _, v in ipairs(conf.header_names) do
+    local token_header = request_headers[v]
+    if token_header then
```
this value `token_header` can potentially not be a string, but a table. This happens when the same header occurs multiple times.

Add something like:
```lua
if type(token_header) == "table" then
  token_header = token_header[1]
end
```

would be nice to add a regression test for this as well (not failing on duplicate headers)

raelmax replied:
@Tieske, sorry for the delay. I add the test and fix the case with a table instead a string. Should I reopen the merge request and close this one?

Tieske commented:
Hmmm just noticed CI seems to be gone from this PR... Maybe better to indeed close this and open a new PR. (not sure that was the cause, but can't think of anything else)

raelmax commented:
Should I close this PR and open another one?

thibaultcha commented:
@raelmax Please keep the same PR.m, but rebase your patch on top of the next branch and force push your changes.

raelmax commented:
Hey guys, any update about the merge of this pull request? =)

thibaultcha commented:
@raelmax Hi,

At the moment we are busy with the 1.0 release, and we are delaying all unrelated work (including feature PRs). Sorry about the delay, once 1.0 out, we'll get back to this PR. Thanks.

oovolkov commented:
Hi guys.
Are there any plans to move forward with this PR.
This feature is highly required.
Thank you in advance.

hishamhm commented on kong/plugins/jwt/handler.lua:
```diff
@@ -41,20 +41,28 @@ local function retrieve_token(conf)
+      end
+      local iterator, iter_err = re_gmatch(token_header, "\\s*[Bb]earer\\s+(.+)")
+      if not iterator then
+        ngx.log(ngx.ERR, "[jwt]: ", iter_err)
```
needs to be replaced with `kong.log.err`

hishamhm commented on kong/plugins/jwt/handler.lua:
```diff
@@ -41,20 +41,28 @@ local function retrieve_token(conf)
-    end
+      local m, err = iterator()
+      if err then
+        ngx.log(ngx.ERR, "[jwt]: ", err)
```
needs to be replaced with `kong.log.err`

oovolkov replied:
Hi guys
Who can fix this PR?
Thank you in advance

hishamhm reviewed (commented):
Hi @raelmax, sorry for not getting back to this! The PR currently has conflicts, but I think a simple rebase should fix it and get it in a approvable/mergeable state!

Tieske committed:
fix review comments

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/kong?pullRequest=3668) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/kong?pullRequest=3668) before we can accept your contribution.<br/><hr/>**Rael Max** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/kong?pullRequest=3668) it.</sub>

Tieske commented:
replaced by #4757

gszr commented:
@raelmax, Thank you for your contribution and apologies for the delay on our end. It has been merged (thanks @Tieske) and will be in Kong 1.3.
